File: src/heap/lc_1046_last_stone_weight/last_stone_weight.py
import heapq
class Solution:
    def lastStoneWeight(self, stones: list[int]) -> int:
        heap = []
        heapq.heapify(heap)
        for stone in stones:
            heapq.heappush(heap, -stone)
        
        while len(heap) > 1:
            x = heapq.heappop(heap)
            y = heapq.heappop(heap)
            if x != y:
                heapq.heappush(heap, x -y)
            else:
                if not heap:
                    return 0
            
        if not heap:
            return 0
        return -heap[0]
        

File: src/heap/lc_1046_last_stone_weight/last_stone_weight.java
import java.util.Comparator;
import java.util.PriorityQueue;

class Solution {
    final PriorityQueue<Integer> pq = new PriorityQueue<>(
        Comparator.reverseOrder()
    ); //max heap
    public int lastStoneWeight(int[] stones) {
        for(int i = 0; i < stones.length; i++) {
            pq.offer(stones[i]);
        }
        while(pq.size() > 1){
            int y = pq.remove();
            int x = pq.remove();
            if (y > x) {
                y = y - x;
                pq.offer(y);
            }
        }
        return pq.size() == 1 ? pq.remove() : 0;
    }
}